src/core/eme/find_key_system.ts

```diff
@@ -260,93 +239,114 @@ function buildKeySystemConfigurations(
  *   - throw if no  compatible key system has been found.
  *
  * @param {Array.<Object>} keySystems - The keySystems you want to test.
+ * @param {Object} currentMediaKeysInfos
  * @returns {Observable}
  */
-function findCompatibleKeySystem(
-  keySystems: IKeySystemOption[],
-  instanceInfos: IInstanceInfo
-) : Observable<IKeySystemAccessInfos> {
-  // Fast way to find a compatible keySystem if the currently loaded
-  // one as exactly the same compatibility options.
-  const cachedKeySystemAccess =
-    getCachedKeySystemAccess(keySystems, instanceInfos);
-  if (cachedKeySystemAccess) {
-    log.debug("eme: found compatible keySystem quickly", cachedKeySystemAccess);
-    return Observable.of(cachedKeySystemAccess);
-  }
-
-  /**
-   * Array of set keySystems for this content.
-   * Each item of this array is an object containing the following keys:
-   *   - keyName {string}: keySystem canonical name (e.g. "widevine")
-   *   - keyType {string}: keySystem type (e.g. "com.widevine.alpha")
-   *   - keySystem {Object}: the original keySystem object
-   * @type {Array.<Object>}
-   */
-  const keySystemsType: IKeySystemType[] = keySystems.reduce(
-    (arr: IKeySystemType[], keySystem) => {
-
-      let ksType;
-
-      if (EME_KEY_SYSTEMS[keySystem.type]) {
-        ksType = EME_KEY_SYSTEMS[keySystem.type].map((keyType) => {
-          const keyName = keySystem.type;
-          return { keyName, keyType, keySystem };
-        }
-        );
-      }
-      else {
-        const keyName = findKeySystemCanonicalName(keySystem.type) || "";
-        const keyType = keySystem.type;
-        ksType = [{ keyName, keyType, keySystem }];
+export default function getMediaKeySystemAccess(
+  mediaElement : HTMLMediaElement,
+  keySystemsConfigs: IKeySystemOption[],
+  currentMediaKeysInfos: MediaKeysInfosStore
+) : Observable<IFoundMediaKeySystemAccessEvent> {
+  return Observable.defer(() => {
+    const currentState = currentMediaKeysInfos.getState(mediaElement);
+    if (currentState) {
+      // Fast way to find a compatible keySystem if the currently loaded
+      // one as exactly the same compatibility options.
+      const cachedKeySystemAccess = checkCachedMediaKeySystemAccess(
+        keySystemsConfigs,
+        currentState.mediaKeySystemAccess,
+        currentState.keySystemOptions
+      );
+      if (cachedKeySystemAccess) {
+        log.debug("eme: found cached compatible keySystem", cachedKeySystemAccess);
+        return Observable.of({
+          type: "reuse-media-key-system-access" as "reuse-media-key-system-access",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=222&issues=AWNklCZWXOBV_YrhYTpf&open=AWNklCZWXOBV_YrhYTpf)

[ISSUE_KEY]: AWNklCZWXOBV_YrhYTpf